modules/raft/src/test/java/org/apache/ignite/raft/jraft/util/concurrent/ConcurrentLinkedLifoByteBufferCollectorPoolTest.java

```diff
@@ -0,0 +1,191 @@
+
+        collectorPool.release(collector0);
+        collectorPool.release(collector1);
+
+        Set<ByteBufferCollector> borrows = ConcurrentHashMap.newKeySet();
+
+        IgniteTestUtils.runRace(
+                () -> borrows.add(collectorPool.borrow()),
+                () -> borrows.add(collectorPool.borrow())
+        );
+
+        assertNull(collectorPool.borrow());
+
+        assertThat(borrows, hasSize(2));
+        assertThat(borrows, hasItems(collector0, collector1));
+    }
+
+    @Test
+    void releaseMoreThanCapacity() {
+        List<ByteBufferCollector> collectors = allocateListCollector(CAPACITY + 5);
+
+        collectors.forEach(collectorPool::release);
+
+        assertEquals(reverse(collectors.subList(0, CAPACITY)), borrowAll());
+    }
+
+    @RepeatedTest(10)
+    void releaseConcurrent() {
+        ByteBufferCollector collector0 = allocateCollector();
+        ByteBufferCollector collector1 = allocateCollector();
+
+        IgniteTestUtils.runRace(
+                () -> collectorPool.release(collector0),
+                () -> collectorPool.release(collector1)
+        );
+
+        List<ByteBufferCollector> borrowAll = borrowAll();
+        assertThat(borrowAll, hasSize(2));
+        assertThat(borrowAll, hasItems(collector0, collector1));
+    }
+
+    @RepeatedTest(10)
+    void borrowReleaseConcurrent() {
+        ByteBufferCollector collector0 = allocateCollector();
+        ByteBufferCollector collector1 = allocateCollector();
+        ByteBufferCollector collector2 = allocateCollector();
+        ByteBufferCollector collector3 = allocateCollector();
+        
+        collectorPool.release(collector0);
+        collectorPool.release(collector1);
+
+        Set<ByteBufferCollector> borrows = ConcurrentHashMap.newKeySet();
+        
+        IgniteTestUtils.runRace(
+                () -> collectorPool.release(collector2),
+                () -> collectorPool.release(collector3),
+                () -> borrows.add(collectorPool.borrow()),
+                () -> borrows.add(collectorPool.borrow())
+        );
+
+        List<ByteBufferCollector> remainingInCollectorPool = borrowAll();
+        
+        assertThat(remainingInCollectorPool, hasSize(2));
+        assertThat(borrows, hasSize(2));
+        
+        assertThat(remainingInCollectorPool, not(hasItems(borrows.toArray(ByteBufferCollector[]::new))));
+    }
+
+    private List<ByteBufferCollector> borrowAll() {
+        var res = new ArrayList<ByteBufferCollector>();
+
+        while (true) {
+            ByteBufferCollector collector = collectorPool.borrow();
+
+            if (collector == null) {
+                break;
+            } else {
+                res.add(collector);
+            }
+        }
+
+        return res;
+    }
+
+    private static ByteBufferCollector allocateCollector() {
+        return ByteBufferCollector.allocate(32);
+    }
+
+    private static List<ByteBufferCollector> allocateListCollector(int size) {
```

Comment:
```suggestion
    private static List<ByteBufferCollector> allocateCollectors(int size) {
```
Or `allocateListOfCollectors()`?